bridges/otellogr/logsink.go

```diff
@@ -0,0 +1,403 @@
+//     set.
+//   - PC is dropped.
+//   - KeyAndValues are transformed and set as Attributes.
+//   - Error is always logged as an additional attribute with the key "err" and
+//     with the severity [log.SeverityError].
+//   - Name is logged as an additional attribute with the key "logger".
+//   - Context is not propagated to the OpenTelemetry log record, as logr does
+//     not provide a way to access it.
+//
+// The Level is transformed by using the [WithLevelSeverity] option. If this is
+// not provided it would default to a function that adds an offset to the logr
+// such that [logr.Info] is transformed to [log.SeverityInfo]. For example:
+//
+//   - [logr.Info] is transformed to [log.SeverityInfo].
+//   - [logr.V(0)] is transformed to [log.SeverityInfo].
+//   - [logr.V(1)] is transformed to [log.SeverityInfo2].
+//   - [logr.V(2)] is transformed to [log.SeverityInfo3].
+//   - ...
+//   - [logr.V(15)] is transformed to [log.SeverityFatal4].
+//   - [logr.Error] is transformed to [log.SeverityError].
+//
+// KeysAndValues values are transformed based on their type. The following types are
+// supported:
+//
+//   - [bool] are transformed to [log.BoolValue].
+//   - [string] are transformed to [log.StringValue].
+//   - [int], [int8], [int16], [int32], [int64] are transformed to
+//     [log.Int64Value].
+//   - [uint], [uint8], [uint16], [uint32], [uint64], [uintptr] are transformed
+//     to [log.Int64Value] or [log.StringValue] if the value is too large.
+//   - [float32], [float64] are transformed to [log.Float64Value].
+//   - [time.Duration] are transformed to [log.Int64Value] with the nanoseconds.
+//   - [complex64], [complex128] are transformed to [log.StringValue] with the
+//   - [time.Time] are transformed to [log.Int64Value] with the nanoseconds.
+//   - [[]byte] are transformed to [log.BytesValue].
+//   - [error] are transformed to [log.StringValue] with the error message.
+//   - [nil] are transformed to an empty [log.Value].
+//   - [struct] are transformed to [log.StringValue] with the struct fields.
+//   - [slice], [array] are transformed to [log.SliceValue] with the elements.
+//   - [map] are transformed to [log.MapValue] with the key-value pairs.
+//   - [pointer], [interface] are transformed to the dereferenced value.
+//
+// [OpenTelemetry]: https://opentelemetry.io/docs/concepts/signals/logs/
+package otellogr // import "go.opentelemetry.io/contrib/bridges/otellogr"
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"reflect"
+	"strconv"
+	"time"
+
+	"github.com/go-logr/logr"
+
+	"go.opentelemetry.io/otel/log"
+	"go.opentelemetry.io/otel/log/global"
+	"go.opentelemetry.io/otel/sdk/instrumentation"
+)
+
+const (
+	bridgeName = "go.opentelemetry.io/contrib/bridges/otellogr"
+	// nameKey is used to log the `WithName` values as an additional attribute.
+	nameKey = "logger"
+	// errKey is used to log the error parameter of Error as an additional attribute.
+	errKey = "err"
```

Comment:
We should use semantic conventions for this, `err` is too go-specific.
Logs have the notion of [exception attributes](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/exceptions/exceptions-logs.md), but not errors. Maybe that's what we should rely on?

Reply:
Actually, thinking about it, I wonder if we shouldn't suggest adding an `error` namespace to semantic conventions.
I have started a semantic conventions discussion about this: https://github.com/open-telemetry/semantic-conventions/issues/921